File: fixtures/scale/java/large/src/cryptobench/ciphers/speck/SpeckRounds.java
package cryptobench.ciphers.speck;

/** Speck64-style ARX rounds: rotate, add, xor. */
class SpeckRounds {
    static long encryptBlock(long block, SpeckKey key) {
        int v0 = (int) (block >>> 32);
        int v1 = (int) block;
        for (int r = 0; r < 27; r++) {
            v0 = (Integer.rotateRight(v0, 8) + v1) ^ key.k(r & 3);
            v1 = Integer.rotateLeft(v1, 3) ^ v0;
        }
        return pack(v0, v1);
    }

    static long decryptBlock(long block, SpeckKey key) {
        int v0 = (int) (block >>> 32);
        int v1 = (int) block;
        for (int r = 27 - 1; r >= 0; r--) {
            v1 = Integer.rotateRight(v1 ^ v0, 3);
            v0 = Integer.rotateLeft((v0 ^ key.k(r & 3)) - v1, 8);
        }
        return pack(v0, v1);
    }

    private static long pack(int v0, int v1) {
        return ((long) v0 << 32) | (v1 & 0xFFFFFFFFL);
    }
}

File: fixtures/scale/dart/large/lib/ciphers/speck/speck_rounds.dart
import 'speck_key.dart';

/// Speck32-style ARX rounds: rotate, add, xor.
int encryptBlock(int block, SpeckKey key) {
  var v0 = (block >> 32) & 0xFFFFFFFF;
  var v1 = block & 0xFFFFFFFF;
  for (var r = 0; r < 27; r++) {
    v0 = (((v0 >> 8 | v0 << 24) & 0xFFFFFFFF) + v1) ^ key.k(r & 3); v0 &= 0xFFFFFFFF;
    v1 = ((v1 << 3 | v1 >> 29) & 0xFFFFFFFF) ^ v0;
  }
  return _pack(v0, v1);
}

int decryptBlock(int block, SpeckKey key) {
  var v0 = (block >> 32) & 0xFFFFFFFF;
  var v1 = block & 0xFFFFFFFF;
  for (var r = 27 - 1; r >= 0; r--) {
    v1 = ((v1 ^ v0) >> 3 | (v1 ^ v0) << 29) & 0xFFFFFFFF;
    v0 = (((v0 ^ key.k(r & 3)) - v1) & 0xFFFFFFFF); v0 = (v0 << 8 | v0 >> 24) & 0xFFFFFFFF;
  }
  return _pack(v0, v1);
}

int _pack(int v0, int v1) => ((v0 & 0xFFFFFFFF) << 32) | (v1 & 0xFFFFFFFF);
